### ISSUE TITLE
Recursive generic type parameters can give confusing error messages

### ISSUE THREAD

**Issue by ltratt** [rust-lang/rust#53191]
If I have code like this:

```rust
struct S<T> {
    parent: Option<Box<S<T>>>
}
```

I get the following compiler error message:

```
error[E0392]: parameter `T` is never used
  --> t.rs:12:10
   |
12 | struct S<T> {
   |          ^ unused type parameter
   |
   = help: consider removing `T` or using a marker such as `std::marker::PhantomData`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0392`.
```

The error message is, IMHO, technically correct, but not necessarily helpful. Now, in this particular example, it's relatively obvious what the mistake is, but I stumbled across it while doing a refactoring that tweaked a complex `enum` -- I was baffled for a while :) My intuitive explanation is that although I'm "using" `T` by passing it recursively to `S` I'm never "consuming" `T` by attaching it to a concrete type.

I wonder if E0392 might want to be special cased / reworded to cover this particular case? I appreciate it's not a common one, but I'm probably not the first person to do something silly like this.

**Comment by jberkenbilt** [comment]
I'm brand new to rust and ran into this while trying to reproduce another issue. I actually don't understand why this is silly or why the error applies. Without seeing this, I generated exactly the same code except my struct was called `A`. :-) I notice that this assertion passes, so I don't see exactly why it's an error without the phantoms.

```rust
use std::marker::PhantomData;
struct A<T> {
    x: Option<Box<A<T>>>,
    _z: PhantomData<T>,
}

fn main() {
    let a = A::<i32> {
        x: None,
        _z: PhantomData
    };
    assert!(a.x.is_none());
}
```

**Comment by peter-kehl** [comment]
If you need recursive generic type parameters without `PhantomData`, and you're happy with `nightly`-only, see #108725. However, we don't know yet whether that is the intended behavior or not, so unsure for how long it will be available.

**Comment by workingjubilee** [comment]
I am usually an advocate for experimenting with nightly features, even "in production", but that is contingent on the fact that most nightly features are actually quite sound. I strongly advise against using an incomplete feature for any practical code of significant use.
